{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2cf05326-032a-43e9-92a7-f2e66ee0f566",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import gzip\n",
    "from os import path as ospath\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import random\n",
    "import pprint\n",
    "from src.utils import write_to_json\n",
    "terms_fp = './downloads/input/go_term_info.json'\n",
    "genes_fp = './downloads/input/bp_module_gene_info.json'\n",
    "term_dispositions_fp = './downloads/input/term_dispositions.json'\n",
    "ibd_modules_fp = './downloads/input/ibd_modules_organized.json'\n",
    "\n",
    "clean_ibd_modules_fp = './downloads/input/clean_ibd_modules.json'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d37cf254-39c6-4cfc-ab16-31fca403096e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_pd_row(df, k):\n",
    "    return dict(df.loc[k].dropna())\n",
    "\n",
    "def spread_terms(df, terms):\n",
    "    return [get_pd_row(df, term) for term in terms]\n",
    "\n",
    "def get_pd_row_key(df, k):\n",
    "    try:\n",
    "        return dict(df.loc[k].dropna())\n",
    "    except KeyError:\n",
    "        return None\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "12a03223-af61-4e2b-9b22-d84a9d5d54da",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>term_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GO:0042552</th>\n",
       "      <td>GO:0042552</td>\n",
       "      <td>myelination</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0045065</th>\n",
       "      <td>GO:0045065</td>\n",
       "      <td>cytotoxic T cell differentiation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0001992</th>\n",
       "      <td>GO:0001992</td>\n",
       "      <td>regulation of systemic arterial blood pressure...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0048247</th>\n",
       "      <td>GO:0048247</td>\n",
       "      <td>lymphocyte chemotaxis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0045954</th>\n",
       "      <td>GO:0045954</td>\n",
       "      <td>positive regulation of natural killer cell med...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OTHER:0008</th>\n",
       "      <td>OTHER:0008</td>\n",
       "      <td>other circadian rhythm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OTHER:0009</th>\n",
       "      <td>OTHER:0009</td>\n",
       "      <td>other metabolic process</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OTHER:0010</th>\n",
       "      <td>OTHER:0010</td>\n",
       "      <td>other system process</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OTHER:0011</th>\n",
       "      <td>OTHER:0011</td>\n",
       "      <td>other cellular process</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OTHER:0012</th>\n",
       "      <td>OTHER:0012</td>\n",
       "      <td>other anatomical structure development</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2015 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    id                                              label\n",
       "term_id                                                                  \n",
       "GO:0042552  GO:0042552                                        myelination\n",
       "GO:0045065  GO:0045065                   cytotoxic T cell differentiation\n",
       "GO:0001992  GO:0001992  regulation of systemic arterial blood pressure...\n",
       "GO:0048247  GO:0048247                              lymphocyte chemotaxis\n",
       "GO:0045954  GO:0045954  positive regulation of natural killer cell med...\n",
       "...                ...                                                ...\n",
       "OTHER:0008  OTHER:0008                             other circadian rhythm\n",
       "OTHER:0009  OTHER:0009                            other metabolic process\n",
       "OTHER:0010  OTHER:0010                               other system process\n",
       "OTHER:0011  OTHER:0011                             other cellular process\n",
       "OTHER:0012  OTHER:0012             other anatomical structure development\n",
       "\n",
       "[2015 rows x 2 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "terms_df = pd.read_json(terms_fp)\n",
    "terms_df = terms_df.set_index('term_id', drop=False)\n",
    "terms_df = terms_df.rename(columns={'term_id': 'id', 'term_label': 'label'})\n",
    "terms_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0045516e-26b6-436a-8750-7fc0d74a8fdd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>disposition</th>\n",
       "      <th>disposition_target_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>term_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GO:0001818</th>\n",
       "      <td>GO:0001818</td>\n",
       "      <td>negative</td>\n",
       "      <td>GO:0001816</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0001915</th>\n",
       "      <td>GO:0001915</td>\n",
       "      <td>negative</td>\n",
       "      <td>GO:0001913</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0001916</th>\n",
       "      <td>GO:0001916</td>\n",
       "      <td>positive</td>\n",
       "      <td>GO:0001913</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0001938</th>\n",
       "      <td>GO:0001938</td>\n",
       "      <td>positive</td>\n",
       "      <td>GO:0001935</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0001953</th>\n",
       "      <td>GO:0001953</td>\n",
       "      <td>negative</td>\n",
       "      <td>GO:0007160</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:1904262</th>\n",
       "      <td>GO:1904262</td>\n",
       "      <td>negative</td>\n",
       "      <td>GO:0038202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:1904263</th>\n",
       "      <td>GO:1904263</td>\n",
       "      <td>positive</td>\n",
       "      <td>GO:0038202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:1903036</th>\n",
       "      <td>GO:1903036</td>\n",
       "      <td>positive</td>\n",
       "      <td>GO:0009611</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:1903070</th>\n",
       "      <td>GO:1903070</td>\n",
       "      <td>negative</td>\n",
       "      <td>GO:0030433</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:1903142</th>\n",
       "      <td>GO:1903142</td>\n",
       "      <td>positive</td>\n",
       "      <td>GO:0061028</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>317 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    id disposition disposition_target_id\n",
       "term_id                                                 \n",
       "GO:0001818  GO:0001818    negative            GO:0001816\n",
       "GO:0001915  GO:0001915    negative            GO:0001913\n",
       "GO:0001916  GO:0001916    positive            GO:0001913\n",
       "GO:0001938  GO:0001938    positive            GO:0001935\n",
       "GO:0001953  GO:0001953    negative            GO:0007160\n",
       "...                ...         ...                   ...\n",
       "GO:1904262  GO:1904262    negative            GO:0038202\n",
       "GO:1904263  GO:1904263    positive            GO:0038202\n",
       "GO:1903036  GO:1903036    positive            GO:0009611\n",
       "GO:1903070  GO:1903070    negative            GO:0030433\n",
       "GO:1903142  GO:1903142    positive            GO:0061028\n",
       "\n",
       "[317 rows x 3 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "term_dispositions_df = pd.read_json(term_dispositions_fp)\n",
    "term_dispositions_df = term_dispositions_df.set_index('term_id', drop=False)\n",
    "term_dispositions_df = term_dispositions_df.rename(columns={'term_id': 'id', 'affected_term_id': 'disposition_target_id'})\n",
    "\n",
    "term_dispositions_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "93aeb04c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>taxon_id</th>\n",
       "      <th>panther_family</th>\n",
       "      <th>long_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>UniProtKB:Q8IWV8</th>\n",
       "      <td>UniProtKB:Q8IWV8</td>\n",
       "      <td>UBR2</td>\n",
       "      <td>E3 ubiquitin-protein ligase UBR2</td>\n",
       "      <td>9606</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UniProtKB:Q96D42</th>\n",
       "      <td>UniProtKB:Q96D42</td>\n",
       "      <td>HAVCR1</td>\n",
       "      <td>Hepatitis A virus cellular receptor 1</td>\n",
       "      <td>9606</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UniProtKB:Q96RW7</th>\n",
       "      <td>UniProtKB:Q96RW7</td>\n",
       "      <td>HMCN1</td>\n",
       "      <td>Hemicentin-1</td>\n",
       "      <td>9606</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UniProtKB:P37059</th>\n",
       "      <td>UniProtKB:P37059</td>\n",
       "      <td>HSD17B2</td>\n",
       "      <td>17-beta-hydroxysteroid dehydrogenase type 2</td>\n",
       "      <td>9606</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UniProtKB:Q8WXG6</th>\n",
       "      <td>UniProtKB:Q8WXG6</td>\n",
       "      <td>MADD</td>\n",
       "      <td>MAP kinase-activating death domain protein</td>\n",
       "      <td>9606</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UniProtKB:Q9H3U5</th>\n",
       "      <td>UniProtKB:Q9H3U5</td>\n",
       "      <td>MFSD1</td>\n",
       "      <td>Major facilitator superfamily domain-containin...</td>\n",
       "      <td>9606</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UniProtKB:O94989</th>\n",
       "      <td>UniProtKB:O94989</td>\n",
       "      <td>ARHGEF15</td>\n",
       "      <td>Rho guanine nucleotide exchange factor 15</td>\n",
       "      <td>9606</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UniProtKB:O00458</th>\n",
       "      <td>UniProtKB:O00458</td>\n",
       "      <td>IFRD1</td>\n",
       "      <td>Interferon-related developmental regulator 1</td>\n",
       "      <td>9606</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UniProtKB:Q96I25</th>\n",
       "      <td>UniProtKB:Q96I25</td>\n",
       "      <td>RBM17</td>\n",
       "      <td>Splicing factor 45</td>\n",
       "      <td>9606</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UniProtKB:Q6NXT2</th>\n",
       "      <td>UniProtKB:Q6NXT2</td>\n",
       "      <td>H3-5</td>\n",
       "      <td>Histone H3.3C</td>\n",
       "      <td>9606</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16350 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                              gene gene_symbol  \\\n",
       "gene                                             \n",
       "UniProtKB:Q8IWV8  UniProtKB:Q8IWV8        UBR2   \n",
       "UniProtKB:Q96D42  UniProtKB:Q96D42      HAVCR1   \n",
       "UniProtKB:Q96RW7  UniProtKB:Q96RW7       HMCN1   \n",
       "UniProtKB:P37059  UniProtKB:P37059     HSD17B2   \n",
       "UniProtKB:Q8WXG6  UniProtKB:Q8WXG6        MADD   \n",
       "...                            ...         ...   \n",
       "UniProtKB:Q9H3U5  UniProtKB:Q9H3U5       MFSD1   \n",
       "UniProtKB:O94989  UniProtKB:O94989    ARHGEF15   \n",
       "UniProtKB:O00458  UniProtKB:O00458       IFRD1   \n",
       "UniProtKB:Q96I25  UniProtKB:Q96I25       RBM17   \n",
       "UniProtKB:Q6NXT2  UniProtKB:Q6NXT2        H3-5   \n",
       "\n",
       "                                                          gene_name taxon_id  \\\n",
       "gene                                                                           \n",
       "UniProtKB:Q8IWV8                   E3 ubiquitin-protein ligase UBR2     9606   \n",
       "UniProtKB:Q96D42              Hepatitis A virus cellular receptor 1     9606   \n",
       "UniProtKB:Q96RW7                                       Hemicentin-1     9606   \n",
       "UniProtKB:P37059        17-beta-hydroxysteroid dehydrogenase type 2     9606   \n",
       "UniProtKB:Q8WXG6         MAP kinase-activating death domain protein     9606   \n",
       "...                                                             ...      ...   \n",
       "UniProtKB:Q9H3U5  Major facilitator superfamily domain-containin...     9606   \n",
       "UniProtKB:O94989          Rho guanine nucleotide exchange factor 15     9606   \n",
       "UniProtKB:O00458       Interferon-related developmental regulator 1     9606   \n",
       "UniProtKB:Q96I25                                 Splicing factor 45     9606   \n",
       "UniProtKB:Q6NXT2                                      Histone H3.3C     9606   \n",
       "\n",
       "                  panther_family  long_id  \n",
       "gene                                       \n",
       "UniProtKB:Q8IWV8             NaN      NaN  \n",
       "UniProtKB:Q96D42             NaN      NaN  \n",
       "UniProtKB:Q96RW7             NaN      NaN  \n",
       "UniProtKB:P37059             NaN      NaN  \n",
       "UniProtKB:Q8WXG6             NaN      NaN  \n",
       "...                          ...      ...  \n",
       "UniProtKB:Q9H3U5             NaN      NaN  \n",
       "UniProtKB:O94989             NaN      NaN  \n",
       "UniProtKB:O00458             NaN      NaN  \n",
       "UniProtKB:Q96I25             NaN      NaN  \n",
       "UniProtKB:Q6NXT2             NaN      NaN  \n",
       "\n",
       "[16350 rows x 6 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_df = pd.read_json(genes_fp, dtype={'taxon_id':str})\n",
    "genes_df = genes_df.set_index('gene', drop=False)\n",
    "genes_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "39a5d220",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>categories</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GO:0071840</td>\n",
       "      <td>[{'id': 'GO:0034330', 'modules': [{'module_ter...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GO:0051179</td>\n",
       "      <td>[{'id': 'OTHER:0003', 'modules': [{'module_ter...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GO:0050896</td>\n",
       "      <td>[{'id': 'GO:0007165', 'modules': [{'module_ter...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GO:0002376</td>\n",
       "      <td>[{'id': 'OTHER:0005', 'modules': [{'module_ter...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GO:0010817</td>\n",
       "      <td>[{'id': 'OTHER:0006', 'modules': [{'module_ter...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>GO:0042592</td>\n",
       "      <td>[{'id': 'OTHER:0007', 'modules': [{'module_ter...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>GO:0007623</td>\n",
       "      <td>[{'id': 'OTHER:0008', 'modules': [{'module_ter...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>GO:0008152</td>\n",
       "      <td>[{'id': 'GO:0006259', 'modules': [{'module_ter...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>GO:0003008</td>\n",
       "      <td>[{'id': 'GO:0003013', 'modules': [{'module_ter...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>GO:0009987</td>\n",
       "      <td>[{'id': 'GO:0007155', 'modules': [{'module_ter...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>GO:0048856</td>\n",
       "      <td>[{'id': 'OTHER:0012', 'modules': [{'module_ter...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>OTHER:0001</td>\n",
       "      <td>[{'id': 'GO:0007548', 'modules': [{'module_ter...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            id                                         categories\n",
       "0   GO:0071840  [{'id': 'GO:0034330', 'modules': [{'module_ter...\n",
       "1   GO:0051179  [{'id': 'OTHER:0003', 'modules': [{'module_ter...\n",
       "2   GO:0050896  [{'id': 'GO:0007165', 'modules': [{'module_ter...\n",
       "3   GO:0002376  [{'id': 'OTHER:0005', 'modules': [{'module_ter...\n",
       "4   GO:0010817  [{'id': 'OTHER:0006', 'modules': [{'module_ter...\n",
       "5   GO:0042592  [{'id': 'OTHER:0007', 'modules': [{'module_ter...\n",
       "6   GO:0007623  [{'id': 'OTHER:0008', 'modules': [{'module_ter...\n",
       "7   GO:0008152  [{'id': 'GO:0006259', 'modules': [{'module_ter...\n",
       "8   GO:0003008  [{'id': 'GO:0003013', 'modules': [{'module_ter...\n",
       "9   GO:0009987  [{'id': 'GO:0007155', 'modules': [{'module_ter...\n",
       "10  GO:0048856  [{'id': 'OTHER:0012', 'modules': [{'module_ter...\n",
       "11  OTHER:0001  [{'id': 'GO:0007548', 'modules': [{'module_ter..."
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ibd_modules_df = pd.read_json(ibd_modules_fp)\n",
    "\n",
    "ibd_modules_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "beec02a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def enrich_json_data(data, terms_df, genes_df):\n",
    "    for entry in data:\n",
    "        entry['label']= get_pd_row(terms_df, entry['id'])['label']\n",
    "        entry['count']= len(entry[\"categories\"])\n",
    "        for category in entry.get(\"categories\", []):\n",
    "            # Add category label\n",
    "            category_id = category.get(\"id\")\n",
    "            category[\"label\"] = get_pd_row(terms_df, category_id)['label']\n",
    "            category['count']= len(category[\"modules\"])\n",
    "\n",
    "            for module in category.get(\"modules\", []):\n",
    "                module_term_id = module.get(\"module_term\")\n",
    "                module[\"label\"] = terms_df.loc[module_term_id, 'label'] if module_term_id in terms_df.index else None\n",
    "                nodes = module.get(\"nodes\", [])\n",
    "                module['count']= len(nodes)\n",
    "                for node in nodes:\n",
    "                    node['terms'] = [{'term_id': term, 'term_label': terms_df.loc[term, 'label']} for term in node.get('terms', []) if term in terms_df.index]\n",
    "                    node['leaf_genes'] = [genes_df.loc[gene].to_dict() for gene in node.get('leaf_genes', []) if gene in genes_df.index]\n",
    "                   \n",
    "\n",
    "with open(ibd_modules_fp, 'r') as file:\n",
    "    json_data = json.load(file)\n",
    "\n",
    "\n",
    "# Call the function to enrich the JSON data\n",
    "enrich_json_data(json_data, terms_df, genes_df)\n",
    "\n",
    "# Output the enriched JSON data\n",
    "print(json.dumps(json_data[:1], indent=2))\n",
    "\n",
    "#json_str = json.loads(json_data)\n",
    "write_to_json(json_data, ospath.join(clean_ibd_modules_fp), indent=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "2f890bab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>section_id</th>\n",
       "      <th>section_label</th>\n",
       "      <th>category_id</th>\n",
       "      <th>category_label</th>\n",
       "      <th>module_label</th>\n",
       "      <th>module_id</th>\n",
       "      <th>disposition_sources</th>\n",
       "      <th>disposition</th>\n",
       "      <th>disposition_target_id</th>\n",
       "      <th>node_id</th>\n",
       "      <th>node_label</th>\n",
       "      <th>terms</th>\n",
       "      <th>leaf_genes</th>\n",
       "      <th>category_count</th>\n",
       "      <th>module_count</th>\n",
       "      <th>node_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GO:0071840</td>\n",
       "      <td>cellular component organization or biogenesis</td>\n",
       "      <td>GO:0034330</td>\n",
       "      <td>cell junction organization</td>\n",
       "      <td>tight junction assembly</td>\n",
       "      <td>GO:0120192</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PTN007679700</td>\n",
       "      <td>PATJ HOMOLOG</td>\n",
       "      <td>[{'term_id': 'GO:0120192', 'term_label': 'tigh...</td>\n",
       "      <td>[{'gene': 'UniProtKB:O75970', 'gene_symbol': '...</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GO:0071840</td>\n",
       "      <td>cellular component organization or biogenesis</td>\n",
       "      <td>GO:0034330</td>\n",
       "      <td>cell junction organization</td>\n",
       "      <td>tight junction assembly</td>\n",
       "      <td>GO:0120192</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PTN002698193</td>\n",
       "      <td>LIPOLYSIS-STIMULATED LIPOPROTEIN RECEPTOR</td>\n",
       "      <td>[{'term_id': 'GO:1904274', 'term_label': 'tric...</td>\n",
       "      <td>[{'gene': 'UniProtKB:Q86X29', 'gene_symbol': '...</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GO:0071840</td>\n",
       "      <td>cellular component organization or biogenesis</td>\n",
       "      <td>GO:0034330</td>\n",
       "      <td>cell junction organization</td>\n",
       "      <td>tight junction assembly</td>\n",
       "      <td>GO:0120192</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PTN002731222</td>\n",
       "      <td>PROTEIN POF1B</td>\n",
       "      <td>[{'term_id': 'GO:0070830', 'term_label': 'bice...</td>\n",
       "      <td>[{'gene': 'UniProtKB:Q8WVV4', 'gene_symbol': '...</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GO:0071840</td>\n",
       "      <td>cellular component organization or biogenesis</td>\n",
       "      <td>GO:0034330</td>\n",
       "      <td>cell junction organization</td>\n",
       "      <td>tight junction assembly</td>\n",
       "      <td>GO:0120192</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PTN000462700</td>\n",
       "      <td>FERM AND PDZ DOMAIN-CONTAINING PROTEIN 2</td>\n",
       "      <td>[{'term_id': 'GO:0070830', 'term_label': 'bice...</td>\n",
       "      <td>[{'gene': 'UniProtKB:Q68DX3', 'gene_symbol': '...</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GO:0071840</td>\n",
       "      <td>cellular component organization or biogenesis</td>\n",
       "      <td>GO:0034330</td>\n",
       "      <td>cell junction organization</td>\n",
       "      <td>tight junction assembly</td>\n",
       "      <td>GO:0120192</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PTN004555956</td>\n",
       "      <td>MARVEL DOMAIN-CONTAINING PROTEIN 3</td>\n",
       "      <td>[{'term_id': 'GO:0070830', 'term_label': 'bice...</td>\n",
       "      <td>[{'gene': 'UniProtKB:Q96A59', 'gene_symbol': '...</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5564</th>\n",
       "      <td>OTHER:0001</td>\n",
       "      <td>other biological process</td>\n",
       "      <td>GO:0060078</td>\n",
       "      <td>regulation of postsynaptic membrane potential</td>\n",
       "      <td>regulation of postsynaptic membrane potential</td>\n",
       "      <td>GO:0060078</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PTN001806145</td>\n",
       "      <td>PRO-FMRFAMIDE-RELATED NEUROPEPTIDE FF</td>\n",
       "      <td>[{'term_id': 'GO:0060079', 'term_label': 'exci...</td>\n",
       "      <td>[{'gene': 'UniProtKB:O15130', 'gene_symbol': '...</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5565</th>\n",
       "      <td>OTHER:0001</td>\n",
       "      <td>other biological process</td>\n",
       "      <td>GO:0060078</td>\n",
       "      <td>regulation of postsynaptic membrane potential</td>\n",
       "      <td>regulation of postsynaptic membrane potential</td>\n",
       "      <td>GO:0060078</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PTN000436343</td>\n",
       "      <td>GAMMA-AMINOBUTYRIC ACID RECEPTOR ALPHA-LIKE</td>\n",
       "      <td>[{'term_id': 'GO:0060078', 'term_label': 'regu...</td>\n",
       "      <td>[{'gene': 'UniProtKB:P47869', 'gene_symbol': '...</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5566</th>\n",
       "      <td>OTHER:0001</td>\n",
       "      <td>other biological process</td>\n",
       "      <td>GO:0060078</td>\n",
       "      <td>regulation of postsynaptic membrane potential</td>\n",
       "      <td>regulation of postsynaptic membrane potential</td>\n",
       "      <td>GO:0060078</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PTN000751216</td>\n",
       "      <td>NMDA RECEPTOR 2, ISOFORM C</td>\n",
       "      <td>[{'term_id': 'GO:0060079', 'term_label': 'exci...</td>\n",
       "      <td>[{'gene': 'UniProtKB:Q14957', 'gene_symbol': '...</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5567</th>\n",
       "      <td>OTHER:0001</td>\n",
       "      <td>other biological process</td>\n",
       "      <td>GO:0060078</td>\n",
       "      <td>regulation of postsynaptic membrane potential</td>\n",
       "      <td>regulation of postsynaptic membrane potential</td>\n",
       "      <td>GO:0060078</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PTN002823877</td>\n",
       "      <td>INHIBITORY SYNAPTIC FACTOR 2A</td>\n",
       "      <td>[{'term_id': 'GO:0060080', 'term_label': 'inhi...</td>\n",
       "      <td>[{'gene': 'UniProtKB:A6NMK8', 'gene_symbol': '...</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5568</th>\n",
       "      <td>OTHER:0001</td>\n",
       "      <td>other biological process</td>\n",
       "      <td>GO:0060078</td>\n",
       "      <td>regulation of postsynaptic membrane potential</td>\n",
       "      <td>regulation of postsynaptic membrane potential</td>\n",
       "      <td>GO:0060078</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PTN002698114</td>\n",
       "      <td>INHIBITORY SYNAPTIC FACTOR 1</td>\n",
       "      <td>[{'term_id': 'GO:0060080', 'term_label': 'inhi...</td>\n",
       "      <td>[{'gene': 'UniProtKB:Q2T9L4', 'gene_symbol': '...</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5569 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      section_id                                  section_label category_id  \\\n",
       "0     GO:0071840  cellular component organization or biogenesis  GO:0034330   \n",
       "1     GO:0071840  cellular component organization or biogenesis  GO:0034330   \n",
       "2     GO:0071840  cellular component organization or biogenesis  GO:0034330   \n",
       "3     GO:0071840  cellular component organization or biogenesis  GO:0034330   \n",
       "4     GO:0071840  cellular component organization or biogenesis  GO:0034330   \n",
       "...          ...                                            ...         ...   \n",
       "5564  OTHER:0001                       other biological process  GO:0060078   \n",
       "5565  OTHER:0001                       other biological process  GO:0060078   \n",
       "5566  OTHER:0001                       other biological process  GO:0060078   \n",
       "5567  OTHER:0001                       other biological process  GO:0060078   \n",
       "5568  OTHER:0001                       other biological process  GO:0060078   \n",
       "\n",
       "                                     category_label  \\\n",
       "0                        cell junction organization   \n",
       "1                        cell junction organization   \n",
       "2                        cell junction organization   \n",
       "3                        cell junction organization   \n",
       "4                        cell junction organization   \n",
       "...                                             ...   \n",
       "5564  regulation of postsynaptic membrane potential   \n",
       "5565  regulation of postsynaptic membrane potential   \n",
       "5566  regulation of postsynaptic membrane potential   \n",
       "5567  regulation of postsynaptic membrane potential   \n",
       "5568  regulation of postsynaptic membrane potential   \n",
       "\n",
       "                                       module_label   module_id  \\\n",
       "0                           tight junction assembly  GO:0120192   \n",
       "1                           tight junction assembly  GO:0120192   \n",
       "2                           tight junction assembly  GO:0120192   \n",
       "3                           tight junction assembly  GO:0120192   \n",
       "4                           tight junction assembly  GO:0120192   \n",
       "...                                             ...         ...   \n",
       "5564  regulation of postsynaptic membrane potential  GO:0060078   \n",
       "5565  regulation of postsynaptic membrane potential  GO:0060078   \n",
       "5566  regulation of postsynaptic membrane potential  GO:0060078   \n",
       "5567  regulation of postsynaptic membrane potential  GO:0060078   \n",
       "5568  regulation of postsynaptic membrane potential  GO:0060078   \n",
       "\n",
       "     disposition_sources disposition disposition_target_id       node_id  \\\n",
       "0                     []         NaN                   NaN  PTN007679700   \n",
       "1                     []         NaN                   NaN  PTN002698193   \n",
       "2                     []         NaN                   NaN  PTN002731222   \n",
       "3                     []         NaN                   NaN  PTN000462700   \n",
       "4                     []         NaN                   NaN  PTN004555956   \n",
       "...                  ...         ...                   ...           ...   \n",
       "5564                  []         NaN                   NaN  PTN001806145   \n",
       "5565                  []         NaN                   NaN  PTN000436343   \n",
       "5566                  []         NaN                   NaN  PTN000751216   \n",
       "5567                  []         NaN                   NaN  PTN002823877   \n",
       "5568                  []         NaN                   NaN  PTN002698114   \n",
       "\n",
       "                                       node_label  \\\n",
       "0                                    PATJ HOMOLOG   \n",
       "1       LIPOLYSIS-STIMULATED LIPOPROTEIN RECEPTOR   \n",
       "2                                   PROTEIN POF1B   \n",
       "3        FERM AND PDZ DOMAIN-CONTAINING PROTEIN 2   \n",
       "4              MARVEL DOMAIN-CONTAINING PROTEIN 3   \n",
       "...                                           ...   \n",
       "5564        PRO-FMRFAMIDE-RELATED NEUROPEPTIDE FF   \n",
       "5565  GAMMA-AMINOBUTYRIC ACID RECEPTOR ALPHA-LIKE   \n",
       "5566                   NMDA RECEPTOR 2, ISOFORM C   \n",
       "5567                INHIBITORY SYNAPTIC FACTOR 2A   \n",
       "5568                 INHIBITORY SYNAPTIC FACTOR 1   \n",
       "\n",
       "                                                  terms  \\\n",
       "0     [{'term_id': 'GO:0120192', 'term_label': 'tigh...   \n",
       "1     [{'term_id': 'GO:1904274', 'term_label': 'tric...   \n",
       "2     [{'term_id': 'GO:0070830', 'term_label': 'bice...   \n",
       "3     [{'term_id': 'GO:0070830', 'term_label': 'bice...   \n",
       "4     [{'term_id': 'GO:0070830', 'term_label': 'bice...   \n",
       "...                                                 ...   \n",
       "5564  [{'term_id': 'GO:0060079', 'term_label': 'exci...   \n",
       "5565  [{'term_id': 'GO:0060078', 'term_label': 'regu...   \n",
       "5566  [{'term_id': 'GO:0060079', 'term_label': 'exci...   \n",
       "5567  [{'term_id': 'GO:0060080', 'term_label': 'inhi...   \n",
       "5568  [{'term_id': 'GO:0060080', 'term_label': 'inhi...   \n",
       "\n",
       "                                             leaf_genes  category_count  \\\n",
       "0     [{'gene': 'UniProtKB:O75970', 'gene_symbol': '...               7   \n",
       "1     [{'gene': 'UniProtKB:Q86X29', 'gene_symbol': '...               7   \n",
       "2     [{'gene': 'UniProtKB:Q8WVV4', 'gene_symbol': '...               7   \n",
       "3     [{'gene': 'UniProtKB:Q68DX3', 'gene_symbol': '...               7   \n",
       "4     [{'gene': 'UniProtKB:Q96A59', 'gene_symbol': '...               7   \n",
       "...                                                 ...             ...   \n",
       "5564  [{'gene': 'UniProtKB:O15130', 'gene_symbol': '...              12   \n",
       "5565  [{'gene': 'UniProtKB:P47869', 'gene_symbol': '...              12   \n",
       "5566  [{'gene': 'UniProtKB:Q14957', 'gene_symbol': '...              12   \n",
       "5567  [{'gene': 'UniProtKB:A6NMK8', 'gene_symbol': '...              12   \n",
       "5568  [{'gene': 'UniProtKB:Q2T9L4', 'gene_symbol': '...              12   \n",
       "\n",
       "      module_count  node_count  \n",
       "0                7           9  \n",
       "1                7           9  \n",
       "2                7           9  \n",
       "3                7           9  \n",
       "4                7           9  \n",
       "...            ...         ...  \n",
       "5564             1           5  \n",
       "5565             1           5  \n",
       "5566             1           5  \n",
       "5567             1           5  \n",
       "5568             1           5  \n",
       "\n",
       "[5569 rows x 16 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def find_disposition_sources(module_id, df):\n",
    "    matching_terms = df[df['disposition_target_id'] == module_id]\n",
    "    sources = [{'term_id': row['id'], 'disposition': row['disposition']} for index, row in matching_terms.iterrows()]\n",
    "    return sources\n",
    "\n",
    "flattened_data = []\n",
    "\n",
    "for section in ibd_modules_df.itertuples(index=False):\n",
    "    section_id = section.id\n",
    "    category_count= len(section.categories)\n",
    "    for category in section.categories:\n",
    "        category_id = category['id']\n",
    "        module_count = len(category['modules'])\n",
    "        \n",
    "        for module in category['modules']:\n",
    "          \n",
    "            module_id = module['module_term']\n",
    "            disposition_sources = find_disposition_sources(module_id, term_dispositions_df)\n",
    "\n",
    "            node_count = len(module.get(\"nodes\", []))\n",
    "            \n",
    "            for node in module.get('nodes', []):\n",
    "                node_id = node.get('ptn_id')\n",
    "                node_label = node.get('label')\n",
    "                \n",
    "                term_info = [{'term_id': term, 'term_label': terms_df.loc[term, 'label']} for term in node.get('terms', []) if term in terms_df.index]\n",
    "                gene_info = [genes_df.loc[gene].to_dict() for gene in node.get('leaf_genes', []) if gene in genes_df.index]\n",
    "\n",
    "                flattened_data.append({\n",
    "                    'section_id': section_id,\n",
    "                    'category_id': category_id,\n",
    "                    'module_id': module_id,\n",
    "                    'disposition_sources': disposition_sources,\n",
    "                    'node_id': node_id,\n",
    "                    'node_label': node_label,\n",
    "                    'terms': term_info,\n",
    "                    'leaf_genes': gene_info,\n",
    "                    'category_count': category_count,\n",
    "                    'module_count': module_count,\n",
    "                    'node_count': node_count\n",
    "                })\n",
    "\n",
    "flat_df = pd.DataFrame(flattened_data)\n",
    "\n",
    "flat_df = flat_df.merge(terms_df, left_on='section_id', right_on='id', how='left', suffixes=('', '_section')).rename(columns={'label': 'section_label'})\n",
    "flat_df = flat_df.merge(terms_df, left_on='category_id', right_on='id', how='left', suffixes=('', '_category')).rename(columns={'label': 'category_label'})\n",
    "flat_df = flat_df.merge(terms_df, left_on='module_id', right_on='id', how='left', suffixes=('', '_module_term')).rename(columns={'label': 'module_label'})\n",
    "flat_df = flat_df.merge(term_dispositions_df, left_on='module_id', right_on='id', how='left')\n",
    "\n",
    "\n",
    "final_df = flat_df[['section_id',\n",
    "                    'section_label',\n",
    "                    'category_id', \n",
    "                    'category_label',                     \n",
    "                    'module_label',\n",
    "                    'module_id', \n",
    "                    'disposition_sources',\n",
    "                    'disposition',\n",
    "                    'disposition_target_id',\n",
    "                    'node_id', \n",
    "                    'node_label', \n",
    "                    'terms', \n",
    "                    'leaf_genes',\n",
    "                    'category_count',\n",
    "                    'module_count',\n",
    "                    'node_count']]\n",
    "\n",
    "\n",
    "\n",
    "final_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b42bb991",
   "metadata": {},
   "outputs": [],
   "source": [
    "ibd_json = final_df.to_json(orient=\"records\", default_handler=None)\n",
    "json_str = json.loads(ibd_json)\n",
    "write_to_json(json_str, ospath.join(clean_ibd_modules_fp), indent=2)\n",
    "#pprint.pprint(json_str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea65b8fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_df = final_df[final_df['disposition_sources'].apply(lambda x: len(x) > 1)]\n",
    "\n",
    "\n",
    "filtered_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4fd7562",
   "metadata": {},
   "outputs": [],
   "source": [
    "annos_df = pd.read_json(annotations_fp)[:10]\n",
    "annos_df = annos_df.merge(genes_df[\n",
    "    ['taxon_id', \n",
    "     'taxon_label', \n",
    "     'taxon_abbr' , \n",
    "     'coordinates_chr_num',\n",
    "     'coordinates_start',\n",
    "     'coordinates_end',\n",
    "     'coordinates_strand']], how='left', left_on=\"gene\", right_index=True)\n",
    "#anno_json = annos_df.to_json(orient=\"records\", default_handler=None)\n",
    "#json_str = json.loads(anno_json)\n",
    "#write_to_json(json_str, ospath.join(sample_annotations_fp))\n",
    "#pprint.pprint(json_str)\n",
    "annos_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9439c0a7-aec5-499d-9133-f6ecb50f5714",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "\n",
    "annos_df = pd.read_json(annotations_fp)[:100]\n",
    "annos_df = annos_df.merge(genes_df[['gene_symbol',\n",
    "     'gene_name','taxon_id', 'taxon_label',\n",
    "                                    'taxon_abbr', \n",
    "                                    'coordinates_chr_num','coordinates_start','coordinates_end','coordinates_strand']], how='left', left_on=\"gene\", right_index=True)\n",
    "#annos_df['aspect'] = annos_df['term'].apply(lambda x: get_pd_row(terms_df, x)['aspect'])\n",
    "annos_df\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff15c263",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "\n",
    "def count_evidence(evidences):\n",
    "    return len(evidences)\n",
    "\n",
    "def generate_id(row):\n",
    "    return f'{row[\"gene\"]}_{row[\"_id\"]}'\n",
    "\n",
    "def get_groups(evidences):\n",
    "    groups = set()\n",
    "    for evidence in evidences:\n",
    "        for group in evidence['groups']:\n",
    "            groups.add(group)\n",
    "            \n",
    "    return list(groups)\n",
    "    \n",
    "annos_df = pd.read_json(annotations_fp)[:5]\n",
    "annos_df = annos_df.merge(genes_df[\n",
    "    ['gene_symbol',\n",
    "     'gene_name',\n",
    "     'taxon_id', \n",
    "     'taxon_label', \n",
    "     'taxon_abbr' , \n",
    "     'coordinates_chr_num',\n",
    "     'coordinates_start',\n",
    "     'coordinates_end',\n",
    "     'coordinates_strand']], how='left', left_on=\"gene\", right_index=True)\n",
    "annos_df['aspect'] = annos_df['term'].apply(lambda x: get_pd_row(terms_df, x)['aspect'])\n",
    "annos_df['term'] = annos_df['term'].apply(lambda x: get_pd_row(terms_df, x))\n",
    "annos_df['term_type'] = annos_df['term'].apply(lambda x: term_type(x))\n",
    "annos_df['slim_terms'] = annos_df['slim_terms'].apply(lambda x: spread_terms(terms_df, x))\n",
    "annos_df['qualifier'] = annos_df['qualifier'].str.replace('_', ' ')\n",
    "annos_df['evidence'] = annos_df.apply(lambda x: get_evidence(articles_df, genes_df, x),axis=1)\n",
    "annos_df['evidence_type'] = annos_df.apply(lambda x: get_evidence_type(x), axis=1)\n",
    "annos_df['groups'] = annos_df['evidence'].apply(lambda x: get_groups(x))\n",
    "annos_df['evidence_count'] = annos_df['evidence'].apply(lambda x: count_evidence(x))\n",
    "\n",
    "\n",
    "   \n",
    "anno_json = annos_df.to_json(orient=\"records\", default_handler=None)\n",
    "json_str = json.loads(anno_json)\n",
    "#pprint.pprint(json_str)\n",
    "write_to_json(json_str, ospath.join('.', clean_annotations_fp))\n",
    "annos_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "430b8ab8",
   "metadata": {},
   "outputs": [],
   "source": [
    "annos_df['_id'] = annos_df.index\n",
    "annos_df['_id'] = annos_df.apply(lambda x: generate_id(x), axis=1)\n",
    "annos_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad1bf4cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "\n",
    "def term_display_id(term):\n",
    "    return term['id'] if term['id'].startswith(\"GO\") else ''\n",
    "\n",
    "fields=['gene', 'term']\n",
    "export_annos_df = pd.read_json(annotations_fp)[:5]\n",
    "export_annos_df = export_annos_df[fields]\n",
    "export_annos_df = export_annos_df.merge(genes_df[\n",
    "    ['gene_symbol',\n",
    "     'gene_name']], how='left', left_on=\"gene\", right_index=True)\n",
    "export_annos_df['term'] = export_annos_df['term'].apply(lambda x: get_pd_row(terms_df, x))\n",
    "export_annos_df['term_id'] = export_annos_df['term'].apply(lambda x: term_display_id(x))\n",
    "export_annos_df['term_label'] = export_annos_df['term'].apply(lambda x: x['label'])\n",
    "export_annos_df = export_annos_df.drop(columns=['term'])\n",
    "\n",
    "compression_opts = dict(method='zip',\n",
    "                        archive_name='annotations.csv')  \n",
    "export_annos_df.to_csv(ospath.join('.', export_annotations_csv_fp), index=False, compression=compression_opts)  \n",
    "   \n",
    "    \n",
    "export_anno_json = export_annos_df.to_json(orient=\"records\", default_handler=None)\n",
    "json_str = json.loads(export_anno_json)\n",
    "#pprint.pprint(json_str)\n",
    "write_to_json(json_str, ospath.join('.', export_annotations_json_fp), zip=True)\n",
    "\n",
    "\n",
    "    \n",
    "export_annos_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf19e388",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "    \n",
    "old_annos_df = pd.read_json(old_annotations_fp)\n",
    "old_annos_df = old_annos_df.merge(genes_df[\n",
    "    [ 'taxon_id', \n",
    "     'taxon_label', \n",
    "     'taxon_abbr' , \n",
    "     'coordinates_chr_num',\n",
    "     'coordinates_start',\n",
    "     'coordinates_end',\n",
    "     'coordinates_strand']], how='left', left_on=\"gene\", right_index=True)\n",
    "\n",
    "   \n",
    "anno_json = old_annos_df.to_json(orient=\"records\", default_handler=None)\n",
    "json_str = json.loads(anno_json)\n",
    "#pprint.pprint(json_str)\n",
    "write_to_json(json_str, ospath.join('.', old_clean_annotations_fp))\n",
    "old_annos_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa5fd152-d500-4fb8-9e49-d57e7b76077e",
   "metadata": {},
   "outputs": [],
   "source": [
    "annos_df['evidence_type'] = annos_df.apply(lambda x: get_evidence_type(x), axis=1)\n",
    "annos_df[annos_df['evidence_type']=='direct']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9ee6956-3967-4fb7-be70-1a988a74bbe4",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(dict(annos_df.iloc[1]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc5ceba5-52d7-469d-9cd7-ebec7f9dd373",
   "metadata": {},
   "outputs": [],
   "source": [
    "clean_annos_df = pd.read_json(clean_annotations_fp)\n",
    "clean_annos_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8ea5a92",
   "metadata": {},
   "outputs": [],
   "source": [
    "old_clean_annos_df = pd.read_json(old_clean_annotations_fp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b0b1c0eb-0c68-4201-8e2b-63d7ef18c499",
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_terms = terms_df[terms_df['is_goslim']==True]\n",
    "filtered_terms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6698108b-daab-416a-9804-b4ef8e7b6e7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_terms():\n",
    "    results = list()\n",
    "    for n in range(0, len(human_df) -1):\n",
    "        index = random.randrange(0, len(filtered_terms))\n",
    "        results.append(dict(filtered_terms.iloc[index]))\n",
    "    \n",
    "    return results\n",
    "\n",
    "count = 1\n",
    "\n",
    "def add_gene():\n",
    "    return human_df.iloc[count]['gene']\n",
    "\n",
    "    \n",
    "                            \n",
    "#add_terms()\n",
    "\n",
    "#ann2_df = pd.DataFrame.from_dict(add_terms())\n",
    "ann2_df\n",
    "\n",
    "ann3 = human_df.agg('-'.join, axis=1)\n",
    "ann2_df\n",
    "\n",
    "add_gene()\n",
    "ann2_df['gene'] = ann2_df['id'].apply(lambda x: add_gene())\n",
    "ann2_df\n",
    "\n",
    "json_chunk = ann2_df.to_json(orient=\"records\", default_handler=None)\n",
    "json_str = json.loads(json_chunk)\n",
    "\n",
    "write_to_json(json_str, 'out.json')\n",
    "                     "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da44d01a-2a8e-494e-b08b-45a91c7449c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "class CustomJSONizer(json.JSONEncoder):\n",
    "    def default(self, obj):\n",
    "        return super().encode(bool(obj)) \\\n",
    "            if isinstance(obj, np.bool_) \\\n",
    "            else super().default(obj)\n",
    "    \n",
    "def count_unique_terms(annos_df):\n",
    "    terms = set()\n",
    "    for term in list(annos_df['term']):\n",
    "        terms.add(term['id'])\n",
    "    \n",
    "    return list(terms)\n",
    "\n",
    "#count_uniq = count_unique_terms()\n",
    "#write_to_json(count_uniq, filtered_terms_fp, cls=CustomJSONizer)\n",
    "#print(len(count_uniq))\n",
    "\n",
    "def count_unique_slim_terms(annos_df):\n",
    "    terms = set()\n",
    "        \n",
    "    for s_terms in list(annos_df['slim_terms']):\n",
    "        for term in s_terms:\n",
    "            terms.add(term['id'])\n",
    "            \n",
    "    return list(terms)\n",
    "\n",
    "#count_uniq = count_unique_slim_terms()\n",
    "#write_to_json(count_uniq, filtered_slim_terms_fp, cls=CustomJSONizer)\n",
    "#print(count_uniq)\n",
    "def count_unique_refs(annos_df):\n",
    "    refs = set()\n",
    "        \n",
    "    for evidences in list(annos_df['evidence']):\n",
    "        for evidence in evidences:\n",
    "            for ref in evidence['references']:\n",
    "                if ref is not None:\n",
    "                    refs.add(ref['pmid'])\n",
    "            \n",
    "    return list(refs)\n",
    "\n",
    "def count_unique_withs(annos_df):\n",
    "    gene = set()\n",
    "    gene_name=set()\n",
    "        \n",
    "    for evidences in list(annos_df['evidence']):\n",
    "        for evidence in evidences:\n",
    "            gene.add(evidence['with_gene_id']['gene'])\n",
    "            \n",
    "    return list(gene)\n",
    "\n",
    "len(count_unique_refs(clean_annos_df))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "47ca3e96-d46f-48df-8937-e38aefadc48c",
   "metadata": {},
   "outputs": [],
   "source": [
    "scalar_cols = {\n",
    "    'gene', \n",
    "    'gene_symbol',\n",
    "    'gene_name',\n",
    "    'taxon_id', \n",
    "    'taxon_label', \n",
    "    'taxon_abbr', \n",
    "    'coordinates_chr_num', \n",
    "    'coordinates_start', \n",
    "    'coordinates_end', \n",
    "    'aspect', \n",
    "    'evidence_type'\n",
    "}\n",
    "    \n",
    "\n",
    "stats = {k: len(clean_annos_df[k].unique()) for k in scalar_cols }\n",
    "\n",
    "stats['terms']=len(count_unique_terms(clean_annos_df))\n",
    "stats['slim_terms']=len(count_unique_slim_terms(clean_annos_df))\n",
    "stats['references']=len(count_unique_refs(clean_annos_df))\n",
    "stats['with_gene']=len(count_unique_withs(clean_annos_df))\n",
    "stats_df = pd.DataFrame.from_dict(stats, orient ='index', columns=['Unique Count'])\n",
    "stats_df\n",
    "#clean_annos_df['evidence_type'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef470f62",
   "metadata": {},
   "outputs": [],
   "source": [
    "scalar_cols = {\n",
    "    'gene', \n",
    "    'gene_symbol',\n",
    "    'gene_name',\n",
    "    'taxon_id', \n",
    "    'taxon_label', \n",
    "    'taxon_abbr', \n",
    "    'coordinates_chr_num', \n",
    "    'coordinates_start', \n",
    "    'coordinates_end', \n",
    "}\n",
    "    \n",
    "\n",
    "stats = {k: len(old_clean_annos_df[k].unique()) for k in scalar_cols }\n",
    "\n",
    "stats_df = pd.DataFrame.from_dict(stats, orient ='index', columns=['Unique Count'])\n",
    "stats_df\n",
    "#clean_annos_df['evidence_type'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d0fa544",
   "metadata": {},
   "outputs": [],
   "source": [
    "unique_genes = clean_annos_df.drop_duplicates(subset=['gene', 'gene_symbol', 'gene_name'])\n",
    "unique_genes = unique_genes[['gene', 'gene_symbol', 'gene_name']]\n",
    "unique_genes = unique_genes[~unique_genes.duplicated(subset=['gene_symbol'])]\n",
    "duplicate_genes = unique_genes[unique_genes.duplicated(subset=['gene'])]\n",
    "duplicate_genes = duplicate_genes[['gene']]\n",
    "res = unique_genes[unique_genes['gene'].isin(list(duplicate_genes['gene']))]\n",
    "res = res.sort_values(by=['gene'])\n",
    "res.to_csv(duplicate_gene_fp)\n",
    "res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da0ff1cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "unique_genes = clean_annos_df.drop_duplicates(subset=['gene'])\n",
    "unique_genes = unique_genes.sort_values(by=['gene_name'])\n",
    "#unique_genes.to_csv(duplicate_gene_fp)\n",
    "unique_genes = unique_genes[['gene', 'gene_symbol', 'gene_name']]\n",
    "duplicate_genes = unique_genes[unique_genes.duplicated(subset=['gene_name'])]\n",
    "duplicate_genes[duplicate_genes['gene_name']=='Uncharacterized protein (Fragment)']\n",
    "#unique_genes.to_csv('./downloads/genes_symb.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f31240a",
   "metadata": {},
   "outputs": [],
   "source": [
    "unique_genes = old_clean_annos_df.drop_duplicates(subset=['gene'])\n",
    "unique_genes = unique_genes.sort_values(by=['gene_name'])\n",
    "#unique_genes.to_csv(duplicate_gene_fp)\n",
    "unique_genes = unique_genes[['gene', 'gene_symbol', 'gene_name']]\n",
    "duplicate_genes_2 = unique_genes[unique_genes.duplicated(subset=['gene_symbol'])]\n",
    "duplicate_genes_2\n",
    "#unique_genes.to_csv('./downloads/genes_symb.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5eb363f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "res = duplicate_genes[~duplicate_genes['gene_symbol'].isin(list(duplicate_genes_2['gene_symbol']))]\n",
    "res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45d16baf",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_df.loc('UniProtKB:X6R8D5')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e391eaf",
   "metadata": {},
   "outputs": [],
   "source": [
    "clean_annos_df['term_label'] = clean_annos_df['term'].apply(lambda x: x['label'])\n",
    "unknown_cc_df = clean_annos_df[clean_annos_df['term_label']=='Unknown cellular component']\n",
    "unique_genes = unknown_cc_df.drop_duplicates(subset=['gene_name'])\n",
    "unique_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0435d20",
   "metadata": {},
   "outputs": [],
   "source": [
    "clean_annos_df = pd.read_json(clean_annotations_fp)[:100]\n",
    "clean_annos_df =  clean_annos_df.drop(['evidence'], axis=1)\n",
    "clean_annos_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "632360d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def uniquify_term(series):\n",
    "    unique_terms = {}\n",
    "    for item in series:\n",
    "        if isinstance(item, dict):\n",
    "            unique_terms[item['id']] = item\n",
    "    return list(unique_terms.values())\n",
    "\n",
    "def uniquify_slim_terms(series):\n",
    "    unique_terms = {}\n",
    "    for item_list in series:\n",
    "        if isinstance(item_list, list):\n",
    "            for item in item_list:\n",
    "                unique_terms[item['id']] = item\n",
    "    return list(unique_terms.values())\n",
    "\n",
    "genes_df = (\n",
    "    clean_annos_df.groupby('gene')\n",
    "    .agg({\n",
    "        'term': uniquify_term,\n",
    "        'slim_terms': uniquify_slim_terms\n",
    "    })\n",
    "    .reset_index()\n",
    ")\n",
    "\n",
    "anno_json = genes_df.to_json(orient=\"records\", default_handler=None)\n",
    "json_str = json.loads(anno_json)\n",
    "#pprint.pprint(json_str)\n",
    "write_to_json(json_str, ospath.join('.', genes_annotations_fp), indent=2)\n",
    "\n",
    "genes_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed2da52a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def uniquify_term(series, evidence_series):\n",
    "    unique_terms = {}\n",
    "    term_counts = {}\n",
    "    duplicates = []\n",
    "\n",
    "    for idx, item in enumerate(series):\n",
    "        if isinstance(item, dict):\n",
    "            term_id = item['id']\n",
    "            term = item.copy()  \n",
    "            term.pop('is_goslim', None) \n",
    "            term['evidence_type'] = evidence_series.iloc[idx] \n",
    "            \n",
    "            if term_id in term_counts:\n",
    "                term_counts[term_id] += 1\n",
    "                duplicates.append(term)\n",
    "            else:\n",
    "                term_counts[term_id] = 1\n",
    "\n",
    "            unique_terms[term_id] = term\n",
    "            \n",
    "    return (list(unique_terms.values()), duplicates)\n",
    "\n",
    "\n",
    "def uniquify_slim_terms(series, evidence_series):\n",
    "    unique_terms = {}\n",
    "    for idx, item_list in enumerate(series):\n",
    "        if isinstance(item_list, list):\n",
    "            for item in item_list:\n",
    "                term = item.copy() \n",
    "                term.pop('is_goslim', None)\n",
    "                term['evidence_type'] = evidence_series.iloc[idx]  # add the evidence_type\n",
    "                unique_terms[term['id']] = term\n",
    "    return list(unique_terms.values())\n",
    "\n",
    "\n",
    "genes_df = clean_annos_df.groupby('gene').apply(lambda group: pd.Series({\n",
    "    'terms': uniquify_term(group['term'], group['evidence_type'])[0],\n",
    "    'duplicated_terms': uniquify_term(group['term'], group['evidence_type'])[1],  \n",
    "    'slim_terms': uniquify_slim_terms(group['slim_terms'], group['evidence_type'])\n",
    "})).reset_index()\n",
    "\n",
    "anno_json = genes_df.to_json(orient=\"records\", default_handler=None)\n",
    "json_str = json.loads(anno_json)\n",
    "#pprint.pprint(json_str)\n",
    "write_to_json(json_str, ospath.join('.', genes_annotations_fp), indent=2)\n",
    "\n",
    "genes_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d8657ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "def uniquify_term(series, evidence_series):\n",
    "    unique_terms = {}\n",
    "    term_counts = {}\n",
    "\n",
    "    for idx, item in enumerate(series):\n",
    "        if isinstance(item, dict):\n",
    "            term_id = item['id']\n",
    "            term = item.copy()  \n",
    "            term.pop('is_goslim', None) \n",
    "            term['evidence_type'] = evidence_series.iloc[idx]  # add the evidence_type\n",
    "            \n",
    "            if term_id in term_counts:\n",
    "                raise ValueError(f\"Duplicate term found: {term}\")\n",
    "            else:\n",
    "                term_counts[term_id] = 1\n",
    "\n",
    "            unique_terms[term_id] = term\n",
    "            \n",
    "    return list(unique_terms.values())\n",
    "\n",
    "\n",
    "def uniquify_slim_terms(series, evidence_series):\n",
    "    unique_terms = {}\n",
    "    for idx, item_list in enumerate(series):\n",
    "        if isinstance(item_list, list):\n",
    "            for item in item_list:\n",
    "                term = item.copy() \n",
    "                term.pop('is_goslim', None)\n",
    "                term['evidence_type'] = evidence_series.iloc[idx]  # add the evidence_type\n",
    "                unique_terms[term['id']] = term\n",
    "    return list(unique_terms.values())\n",
    "\n",
    "columns_to_extract = [\n",
    "    'gene_name',\n",
    "    'taxon_id',\n",
    "    'taxon_label',\n",
    "    'taxon_abbr',\n",
    "    'panther_family',\n",
    "    'long_id',\n",
    "    'coordinates_chr_num',\n",
    "    'coordinates_start',\n",
    "    'coordinates_end',\n",
    "    'gene_symbol'\n",
    "]\n",
    "\n",
    "def group_terms(group):\n",
    "    unique_terms = uniquify_term(group['term'], group['evidence_type'])\n",
    "    slim_terms = uniquify_slim_terms(group['slim_terms'], group['evidence_type'])\n",
    "    return pd.Series({\n",
    "         **{col: group[col].iloc[0] for col in columns_to_extract},\n",
    "        'terms': unique_terms,\n",
    "        'slim_terms': slim_terms,\n",
    "        'terms_count': len(unique_terms)\n",
    "    })\n",
    "\n",
    "genes_df = clean_annos_df.groupby('gene').apply(group_terms).reset_index()\n",
    "genes_df = genes_df.sort_values(by='terms_count', ascending=False).reset_index(drop=True)\n",
    "\n",
    "\n",
    "anno_json = genes_df.to_json(orient=\"records\", default_handler=None)\n",
    "json_str = json.loads(anno_json)\n",
    "#pprint.pprint(json_str)\n",
    "write_to_json(json_str, ospath.join('.', genes_annotations_fp), indent=2)\n",
    "\n",
    "genes_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13bc326c",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1bc0cf0",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  },
  "vscode": {
   "interpreter": {
    "hash": "e3aeb8d7299acc47bfaa6dcf3b4a74342953842fd9ae362d93334745f6b09739"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
